File: python/394/Decode_String.py
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        while '[' in s:
            s=re.sub(r"(\d+)\[(\w+)\]", lambda m:m.group(2)*int(m.group(1)), s)
        return s

File: cpp/394/Decode_String.cpp
class Solution {
public:
    string decodeString(string s) {
        int i = 0;
        return decodeString(s, i);
    }
    
    string decodeString(string &s, int &i){
        string ret;
        
        while(i<s.size()&&s[i]!=']'){
            if(isdigit(s[i])){
                int k = 0;
                while(i<s.size()&&isdigit(s[i])){
                    k = k*10 + (s[i]-'0');
                    i++;
                }
                i++;
                string tmp = decodeString(s, i);
                i++;
                while(k-->0)
                    ret += tmp;
            }
            else{
                ret += s[i++];
            }
        }
        
        return ret;
    }
};
